pkg/manage/cmd_set_password.go

```diff
@@ -25,63 +26,84 @@ This command sets the password of a user by a given user id.
+	onlyUpdate := cmd.Flags().Bool("update", false, "Do not overwrite an existing password.")
 
-			req := &proto.SetPasswordRequest{
-				UserID:   userID,
-				Password: password,
-			}
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
+		defer cancel()
+
+		service, close, err := Dial(ctx, cfg.Address)
+		if err != nil {
+			return fmt.Errorf("connecting to gRPC server: %w", err)
+		}
+		defer close()
+
+		req := &proto.SetPasswordRequest{
+			UserID:     *userID,
+			Password:   *password,
+			OnlyUpdate: *onlyUpdate,
+		}
 
-			if _, err := service.SetPassword(ctx, req); err != nil {
-				return fmt.Errorf("reset password: %w", err)
+		resp, err := service.SetPassword(ctx, req)
+		if err != nil {
+			return fmt.Errorf("reset password: %w", err)
+		}
+
+		if *onlyUpdate {
+			msg := "Password did already exist."
+			if resp.PasswordSet {
+				msg = "Password was set."
 			}
-			return nil
-		},
+			fmt.Println(msg)
+		}
+		return nil
 	}
 
-	cmd.Flags().Int64VarP(&userID, "user_id", "u", 1, "ID of the user account")
-	cmd.Flags().StringVarP(&password, "password", "p", "admin", "New password for the user")
-
 	return cmd
 }
 
-// SetPassword sets hashes and sets the password
+// SetPassword sets hashes and sets the password.
 func (s *Server) SetPassword(ctx context.Context, in *proto.SetPasswordRequest) (*proto.SetPasswordResponse, error) {
-	waitForService(ctx, s.config.AuthHost, s.config.AuthPort)
-	waitForService(ctx, s.config.DatastoreWriterHost, s.config.DatastoreWriterPort)
+	waitForService(ctx, s.config.AuthURL().Host, s.config.DatastoreWriterURL().Host)
+
+	if in.OnlyUpdate {
+		// Check if the current password exists.
+		waitForService(ctx, s.config.DatastoreReaderURL().Host)
+		key := fmt.Sprintf("user/%d/password", in.UserID)
+		var oldPassword string
+		if err := datastore.Get(ctx, s.config.DatastoreReaderURL().String(), key, &oldPassword); err != nil {
+			return nil, fmt.Errorf("fetching old password: %w", err)
+		}
+
+		if oldPassword != "" {
+			return &proto.SetPasswordResponse{PasswordSet: false}, nil
+		}
+	}
 
-	hash, err := hashPassword(ctx, s.config, in.Password)
+	hash, err := hashPassword(ctx, s.config.AuthURL(), in.Password)
 	if err != nil {
 		return nil, fmt.Errorf("hash password: %w", err)
 	}
 
-	if err := setPassword(ctx, s.config, int(in.UserID), hash); err != nil {
+	if err := setPassword(ctx, s.config.DatastoreWriterURL(), int(in.UserID), hash); err != nil {
 		return nil, fmt.Errorf("set password: %w", err)
 	}
-	return new(proto.SetPasswordResponse), nil
+	return &proto.SetPasswordResponse{PasswordSet: true}, nil
 }
 
 // hashPassword returns the hashed form of password as a JSON.
-func hashPassword(ctx context.Context, cfg *ServerConfig, password string) (string, error) {
+func hashPassword(ctx context.Context, authAddr *url.URL, password string) (string, error) {
```

Comment:
Just for me to learn: Why do you want to use a pointer here?

Reply:
The reason was not this line, but line 82:
```golang
if err := datastore.Get(ctx, s.config.DatastoreReaderURL().String(), key, &oldPassword); err != nil {
	return nil, fmt.Errorf("fetching old password: %w", err)
}
```
It did not work to use `s.config.DatastoreReaderURL().String()` as string argument. But it worked again after changing `DatastoreReaderURL()` to return a pointer of an `url.URL`. I did not put any more though into it and changed any place where it was used.

But since you ask: https://play.golang.org/p/o1XORucNggn